sympy/tensor/array/arrayop.py

```diff
@@ -197,9 +198,20 @@ def derive_by_array(expr, dx):
                 raise ValueError("cannot derive by this array")
 
     if isinstance(expr, array_types):
-        expr = ImmutableDenseNDimArray(expr)
+        if isinstance(expr, NDimArray):
+            expr = expr.as_immutable()
+        else:
+            expr = ImmutableDenseNDimArray(expr)
+
         if isinstance(dx, array_types):
-            new_array = [[y.diff(x) for y in expr] for x in dx]
+            if isinstance(expr, SparseNDimArray):
+                lp = len(expr)
+                new_array = {k + i*lp: expr.diff(x)._sparse_array[k]
+                             for k in expr._sparse_array
+                             for i, x in enumerate(dx)
```

Comment:
why not iterate over the sparse array? You are flattening all members and then filtering with an if on the next line.

Reply:
This is because dx is an arraytype, not exactly a sparse array. Besides, I was allowing flatterning of `dx` because `dx` is supposed to be a sequence of non-zero values because 
```
>>> S.Zero._diff_wrt
False
```
So I think `dx` is not supposed to be a sparse array. Am I misunderstanding something here? Please let me know if so, thank you!!

Reply:
I think I got your point! Working on it.